### PR TITLE
Migrate from MySQLdb to PyMySQL

### LINKED ISSUE #29887: mysql_* modules are implicitly Python3 incompatible
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
* mysql_* module

##### ANSIBLE VERSION
```
ansible 2.3.2.0 / master
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
Python 3 target

##### SUMMARY
mysql_* modules always depend on `MySQLdb`, which does not support Python 3.
This inherently creates incompatibility with Python 3. Raising this as bug per [Python 3 support policy](https://docs.ansible.com/ansible/latest/python_3_support.html).

Possible solutions:
* switch all mysql_* modules to use PyMySQL all the time
* allow all mysql_* modules to use whatever MySQL driver is available (MySQLdb, PyMySQL, etc)

##### STEPS TO REPRODUCE
* use mysql_* module on a target host with Python 3

##### EXPECTED RESULTS
mysql_* modules should be able to use a Python 2/3-compatible drivers, such as PyMySQL

##### ACTUAL RESULTS
```
the python mysqldb module is required
```

### PR DESCRIPTION
**Pull request by DSpeichert**
##### SUMMARY
Fixes #29887 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
mysql,  proxysql

##### ANSIBLE VERSION
devel

##### ADDITIONAL INFORMATION
PyMySQL is a pure Python implementation of MySQL connector. It is widely available for Python 2.7 and 3. This eliminates the need to install MySQL client libs that MySQLdb was using just to use these modules.

### CODE REVIEW

**Comment by mattclay** [comment]
@DSpeichert Thank you for working on adding support for PyMySQL. As long as it maintains full compatibility with MySQLdb it should make a good drop-in replacement. If there are significant API differences between the two dependencies then it may be worth considering releasing new modules (under different names) ported to use PyMySQL and then deprecating the old modules.

One thing to keep in mind when making changes to the existing modules is that the modules will need to maintain backwards compatibility by falling back to MySQLdb if PyMySQL is not available. Doing so will prevent the changes from breaking these modules for existing users who do not have PyMySQL installed.

**Comment by DSpeichert** [comment]
@mattclay I plan on adding the fallback to MySQLdb soon :)

**Comment by bmildren** [comment]
Ahh nice - coincidently I was looking at updating the ProxySQL modules to use this package.  ProxySQL 1.4.8 introduces a change to the sql_mode that leads to an error with the mysqlclient (MySQLdb) connector.  Switching to mysql.connector or pymysql resolves the issue as they don't depend on the mysql client lib.

**Comment by DSpeichert** [comment]
@bmildren Are you saying that with ProxySQL >= 1.4.8 the module is completely broken when using MySQL-python? Some of the integration test targets such as RHEL don't seem to have a package for PyMySQL so they will revert to MySQL-python.

**Comment by bmildren** [comment]
In testing we found the incompatibility is with the MySQL 5.7 client libs and ProxySQL 1.4.8 (and relates to a change that introduced the NO_BACKSLASH_ESCAPES sql_mode).  We found as both the mysql.connector and pymysql connectors were pure python implementations neither were affected.  We haven't found a workaround with MySQL-python yet, but there's still more investigation that can be done.

**Comment by DSpeichert** [comment]
@mattclay Please review.

**Comment by decentral1se** [comment]
Oh, thanks for working on this @DSpeichert! Any news? This change is much needed.

**Comment by DSpeichert** [comment]
@lwm rebased.

File: lib/ansible/module_utils/mysql.py
Comment (decentral1se) [inline lib/ansible/module_utils/mysql.py:76]: Uhh, will this not explode if `mysql_driver` is `None`? Could we make an assertion here? I assume this function is only called internally by the module and therefore we can give any developer a hard stop if they're missing the dependency.

File: lib/ansible/modules/database/mysql/mysql_db.py
Comment (decentral1se) [inline lib/ansible/modules/database/mysql/mysql_db.py]: I think it will be confusing that the `module.fail_json` then says that `MySQL-python` is one requirement you can choose after reading this. It might be useful to say here something to the effect of: 'you are encouraged to use PyMySQL for all the benefits it brings but you can still use MySQL-python but it will be deprecated at some point'. The module is in transition to a better supported dependency and I think it is OK to say that up front.

File: lib/ansible/modules/database/mysql/mysql_db.py
Comment (decentral1se) [inline lib/ansible/modules/database/mysql/mysql_db.py]: Should this maybe be: `"The PyMySQL or MySQL-python (soon to be deprecated) module is required."`

Also see https://github.com/ansible/ansible/pull/40123/files#r217887801.

**Comment by decentral1se** [comment]
> @lwm rebased.

Yuuuusssssss :hand:

I've made a few comments, hope they're useful. Hoping this will get merged ASAP!

**Comment by odyssey4me** [comment]
MySQL-python/MySQLdb1 is now totally unmaintained by the looks of things. It does not work with MariaDB 10.2 - [there's a fix that's been pushed up](https://github.com/farcepest/MySQLdb1/pull/131), but no-one is reviewing. It would be really good to get this patch merged to allow us to switch to using a newer, better maintained library!

**Comment by abadger** [comment]
@maxamillion, would you mind taking a look at the distro compatibility of switching to this package?

It's too late for this to make it into 2.7.0 but I might accept a backport of it for 2.7.1.  The code is straightforward.  The only things I don't know (since I'm not a participant or user in the mysql community): 

* is this the best mysql binding to switch to?  (but if the current binding is not functioning, perhaps anything else [meaning this binding] is preferable?)
* From pypi, it looks like PyMySQL is not working on Python-2.6?  MySQL modules are not core modules but it isn't going to be good if the mysql modules can't continue to work with python-2.6 based distros.... Perhaps the documentation will have to say something like "Python-2.6 with MySQL-python or PyMySQL with more recent Python's"?  @lwm if you'd like to wordsmith that sentiment more appropriately since you know this situation better, I'd appreciate it....

File: lib/ansible/module_utils/mysql.py
Comment (abadger) [reply lib/ansible/module_utils/mysql.py:76]: I think that all the modules should have checked this and called fail_json() if it was None before they call mysql_connect().  There is a fragility in depending on that but it's no different than previously where mysqldb_found was being used for the same purpose.

**Comment by dmsimard** [comment]
@abadger @maxamillion the OpenStack community did the switch from mysqldb to pymysql a while back.
While the use case might not be the same, there was a comprehensive analysis done prior to the switch which you may find helpful: https://wiki.openstack.org/wiki/PyMySQL_evaluation

**Comment by abadger** [comment]
Thanks dmsimard, that document makes a good case that pymysql is a
reasonable choice to move to.

**Comment by decentral1se** [comment]
> @abadger @maxamillion the OpenStack community did the switch from mysqldb to pymysql a while back. While the use case might not be the same, there was a comprehensive analysis done prior to the switch which you may find helpful: https://wiki.openstack.org/wiki/PyMySQL_evaluation

Excellent!


> From pypi, it looks like PyMySQL is not working on Python-2.6? MySQL modules are not core modules but it isn't going to be good if the mysql modules can't continue to work with python-2.6 based distros.... Perhaps the documentation will have to say something like "Python-2.6 with MySQL-python or PyMySQL with more recent Python's"? @lwm if you'd like to wordsmith that sentiment more appropriately since you know this situation better, I'd appreciate it....

That makes sense to me! So, I'll add notes to my previous documentation related comments. We need to make it clear that both dependencies are available based on which python version you need to support.

EDIT: See https://github.com/ansible/ansible/pull/40123#discussion_r217909105 and https://github.com/ansible/ansible/pull/40123#discussion_r217909073.

File: lib/ansible/modules/database/mysql/mysql_db.py
Comment (decentral1se) [reply lib/ansible/modules/database/mysql/mysql_db.py]: Right, so, following https://github.com/ansible/ansible/pull/40123#issuecomment-421655404, we need to make it clear here that if you're on Python 2.X, then you need to continue with `python-mysqldb`/`MySQL-python` and if you're using Python 3.X, you can switch to `PyMySQL`.

File: lib/ansible/modules/database/mysql/mysql_db.py
Comment (decentral1se) [reply lib/ansible/modules/database/mysql/mysql_db.py]: OK, following https://github.com/ansible/ansible/pull/40123#issuecomment-421655404, this should probably be something like: `"The PyMySQL (python 3.X) or MySQL-python (python 2.X) module is required."`. WDYT?

File: lib/ansible/modules/database/mysql/mysql_replication.py
Comment (felixfontein) [inline lib/ansible/modules/database/mysql/mysql_replication.py]: Out of curiosity: why aren't you importing `mysql_driver` from `ansible.module_utils.mysql`? Then you don't need this "figure out what to import" in every module.

File: lib/ansible/modules/database/mysql/mysql_replication.py
Comment (felixfontein) [inline lib/ansible/modules/database/mysql/mysql_replication.py]: This message could also be an imported variable from `ansible.module_utils.mysql` (or these two lines a function in `ansible.module_utils.mysql`), so that when the message ever needs to be adjusted, it can be done in one place.

File: lib/ansible/modules/database/mysql/mysql_db.py
Comment (DSpeichert) [reply lib/ansible/modules/database/mysql/mysql_db.py]: That suggests PyMySQL doesn't work on Python 2.X at all, which in fact (according to PyMySQL) it doesn't work on <=2.6. So maybe: `The PyMySQL (Python >= 2.7) or MySQL-python (Python 2.X) module is required.` ?

Comment (felixfontein) [reply lib/ansible/modules/database/mysql/mysql_db.py]: How about `The PyMySQL (Python 2.7 and Python 3.X) or MySQL-python (Python 2.X) module is required.` to emphasize that only PyMySQL will work fine with Python 3?

Comment (decentral1se) [reply lib/ansible/modules/database/mysql/mysql_db.py]: :+1: :+1: :+1:

File: lib/ansible/modules/database/mysql/mysql_replication.py
Comment (DSpeichert) [reply lib/ansible/modules/database/mysql/mysql_replication.py]: Good idea!

File: lib/ansible/modules/database/mysql/mysql_replication.py
Comment (DSpeichert) [reply lib/ansible/modules/database/mysql/mysql_replication.py]: You're right. I initially wanted to alter as little code as possible but that indeed was very redundant. Thanks for comments.

**Comment by odyssey4me** [comment]
@lwm Given that MySQL-python/MySQLdb1 does not work with MariaDB 10.2+ (at all!), I'd suggest always preferring pymysql if it's present, but falling back to mysql-python, rather than making pymysql the py3 lib and mysql-python to the py2 lib.

**Comment by decentral1se** [comment]
> @lwm Given that MySQL-python/MySQLdb1 does not work with MariaDB 10.2+ (at all!), I'd suggest always preferring pymysql if it's present, but falling back to mysql-python, rather than making pymysql the py3 lib and mysql-python to the py2 lib.

Right, another angle! Thanks for raising.

However, I don't think we're making it a py2/py3 lib thing in the logic, we're just making this clear in the documentation of the failures that a dependency is missing (with python versioning metadata). Both are important to be sure (mysql version + python version). However, I don't think we can 'prefer' on the logic side of this anyway - what appears to be preferred in general may not work for specific use cases in the dark underbelly of the 'installed base' ;).  EDIT: probably a core dev can speak more about this!

I like the simplicity of this patch. I think we could add a further warning to the `notes:` section of the module documentation header to say: `MySQL-python does not support MariaDB 10.2+.`. What do you think?

PS. I have no affiliation to this module at all and I just want this patch to be merged, so I am just banging away at it here. So, please nobody mistake me for someone who can actually influence the direction of this being merged :smile:

**Comment by odyssey4me** [comment]
@lwm Apologies, I hadn't read the contents of the PR. Now that I have it's doing exactly what I think is good. I think it's good as-is actually, the extra note might be nice but I don't think it's essential if we're giving two options. Users will simply use whatever they have/prefer or find to work.

**Comment by DSpeichert** [comment]
As you have both noted by now - it's actually always trying PyMySQL first and then falling back to the old MySQL-python. If you'd like me to add any other notes into documentation, I'll be happy to do so promptly. Otherwise, I'm also eagerly waiting for this to be merged.

File: lib/ansible/utils/module_docs_fragments/mysql.py
Comment (felixfontein) [inline lib/ansible/utils/module_docs_fragments/mysql.py]: I would mention both MySQLdb and PyMySQL here (with `(Python 2.7 and Python 3.X)` after PyMySQL, and `(Python 2.X)` after MySQLdb). This is because in the generated documentation, the requirements section can be found at the very top (before all parameters), and the notes are below the parameters section. So having only PyMySQL listed here will make readers believe that MySQLdb is no longer supported (if they don't continue reading).

**Review by felixfontein — APPROVED** [review]
Looks good to me (up to a potential change how the requirements are listed). I've also tested this with MySQLdb (modules `mysql_user` and `mysql_db`), it worked as expected.

shipit

File: lib/ansible/utils/module_docs_fragments/mysql.py
Comment (DSpeichert) [reply lib/ansible/utils/module_docs_fragments/mysql.py]: I didn't include it to not mislead anyone into believing that BOTH are required at the same time.
I also didn't include both with annotation (e.g. `PyMySQL (Python 2.7 and Python 3.X)`) because I wasn't sure if that wasn't automatically parse by some tools somewhere (slim chance but maybe).
I could add it though.

**Comment by abadger** [comment]
It isn't automatically parsed so you should be okay.

File: lib/ansible/utils/module_docs_fragments/mysql.py
Comment (felixfontein) [reply lib/ansible/utils/module_docs_fragments/mysql.py]: I've used similar constructs for `acme_*` modules and `docker_*` modules, where similar alternatives exist; for `docker_*`, one alternative is no Python module anyway, so if this is ever automatically parsed, there needs to be some extended syntax anyway :)
Anyway, adding "or" is a good idea!

**Comment by decentral1se** [comment]
shipit shipit shipit shipit shipit shipit shipit shipit shipit shipit shipit shipit

**Comment by abadger** [comment]
Could you add a changelog fragment for this as well?  Probably under bugfixes:  so something like this: https://github.com/ansible/ansible/blob/devel/changelogs/fragments/get_url.yaml

You can say something like mysql modules can now use the PyMySQL python library so that they can function on Python 3.

**Comment by abadger** [comment]
I can't see anything else to change once the changelog fragment is added.  Anyone else?

**Comment by DSpeichert** [comment]
@abadger For the changelog, should I repeat the same message for every module that was altered (mysql_db, mysql_user, ..., proxysql_db, ...) ? I am not sure how those "fragments" are later compiled and if they are matched with module names. If not, I'll just add:

```
bugfixes:
- mysql_* - PyMySQL (a pure-Python MySQL driver) is now a preferred dependency, providing support for Python 3.X.
```

**Review by mattclay — APPROVED** [review]

**Comment by mattclay** [comment]
I'd definitely like to see this backported all the way to 2.5 if possible. It will be necessary to continue running the tests on almost any newer distribution.

@nitzmahone Any objections to this being backported to stable-2.5?

**Comment by abadger** [comment]
Your single fragment will do nicely :-)

**Comment by odyssey4me** [comment]
@mattclay Porting this all the way back to stable-2.5 would be ideal for us.

**Comment by abadger** [comment]
merged to devel for the 2.8.0 release.

@DSpeichert are you familiar with making backports?  We've added support for https://pypi.org/project/cherry-picker/ to make opening those easier to do.

**Comment by DSpeichert** [comment]
I will try to submit the backport PRs.

**Pull request merged**
